### PR TITLE
Nb/implement passport authentication +  SignIn + SignUP

### PR DESCRIPTION
**Pull request by NicoBellanich**
In this image, we could see the signUP of a new user , the mutation returns the user with some fields, and the jwt , wich we'll be using in every query to the backend, and autherror wich is a simple message about some handled errors.
image

This picture is about the next step, once we signUP , we could signIN, also this mutation returns username field, jwt and autherror.
image

On this last picture, we could see that if we send current user mutation with the jwt catched before, we get all the information about the user we just created. Also here we can apreciate the encripted password and the generated salt by the backend resolver
image

### CODE REVIEW

File: controllers/userController.js [hunk 1]
```diff
@@ -1,50 +1,77 @@
+import { sign } from "jsonwebtoken";
+import { indexOf } from "lodash";
```
Comment (roke77) [inline controllers/userController.js:2]: Remember to run the linter (`npx standard`) to be warned about this syntax violations, like this `indexOf` that you are not using anymore.

File: controllers/userController.js [hunk 2]
```diff
@@ -1,50 +1,77 @@
...
 
-import * as models from './../models'
-import { sign } from 'jsonwebtoken'
+export const findAll = (parent, args, context, info) => {
 
-export const findAll = () => {
-  return models.User.findAll()
-}
+  if(!context.user){
+    return [{ id : 404, username: "error no logueado" }]
```
Comment (roke77) [inline controllers/userController.js:7]: Use English for all error messages.

File: controllers/userController.js [hunk 3]
```diff
@@ -1,50 +1,77 @@
...
+};
 
-export const signUp = async (obj, { data }) => {
-  const user = await SearchByUsername(data.username)
+export const signUp = async (obj, { data }, context) => {
+  const user = await SearchByUsername(data.username, context);
 
   if (user) {
-    return { user: null, accessToken: null, authError: 'usuario ya creado' }
+    return { user: null, accessToken: null, authError: "Usuario ya creado" };
```
Comment (roke77) [inline controllers/userController.js:34]: English here too

File: controllers/userController.js [hunk 4]
```diff
@@ -1,50 +1,77 @@
...
   } else {
-    const newUser = await models.User.create(data)
-    const { id, username } = newUser
-    const accessToken = sign({ id, username }, 'secret', { expiresIn: '10d' })
-    return { user: newUser, jwt: accessToken, authError: null }
+    const newUser = await context.models.User.create(data);
+    const { id, username } = newUser;
+    const accessToken = await sign({ id, username }, "secret", {
```
Comment (roke77) [inline controllers/userController.js:38]: Your `secret` should be secret. Let's explore how to use the `dotenv` and hide it behind a local environment variable.
Also, let's talk on Slack about using the JWT `subject` parameter.

File: controllers/userController.js [hunk 5]
```diff
@@ -1,50 +1,77 @@
...
+      expiresIn: "10d",
+    });
+    return {
```
Comment (roke77) [inline controllers/userController.js:41]: Always leave a line space on top of any `return` statement.

File: controllers/userController.js [hunk 6]
```diff
@@ -1,50 +1,77 @@
...
+  // if (user && user.password === data.password) {
+  if (user && user.passwordMatches(data.password)) {
+    const { id, username } = user;
+    const accessToken = sign({ id, username }, "secret", { expiresIn: "10d" });
+    return {
+      user: user,
+      jwt: accessToken,
+      authError: null,
+    };
+  }
```
Comment (roke77) [inline controllers/userController.js:61]: I'm not sure if this is a copy & paste, but you should only use `passwordMatches` once to validate the password like this:

```suggestion
  if (!user) return { authError: 'User not found' }
  
  if (!user.passwordMatches(data.password)) return { authError: 'Wrong password' }
  
    const { id, username } = user;
    const jwt = sign({ id, username }, "secret", { expiresIn: "10d" });
    
    return { user, jwt };
  }
```
Then the following part from lines 62 to 66 are not necessary.

File: controllers/userController.js [hunk 7]
```diff
@@ -1,50 +1,77 @@
...
   }
-}
+};
+
+export const currentUser = (parent, args, context, info) => {
+  return context.user;
+};
+
+
+const SearchByUsername = (data, context) => {
```
Comment (roke77) [inline controllers/userController.js:74]: Nice to extract the user finding logic to its own function. I suggest to name it `findByUsername({ username })` and to place this function in the User model so you can then call it like `context.models.User.findByUsername({ username: data.username })`

File: models/user.js [hunk 8]
```diff
@@ -1,19 +1,55 @@
...
-      defaultValue: DataTypes.UUIDV4
+  const User = sequelize.define(
+    "User",
+    {
+      id: {
+        type: DataTypes.UUID,
+        primaryKey: true,
+        defaultValue: DataTypes.UUIDV4,
+      },
+      firstname: DataTypes.STRING,
```
Comment (roke77) [inline models/user.js:12]: It's important to be specific on the type configuration for each field. Add the following:
`allowNull: false` for all fields, and `unique: true` for `username`.

File: models/user.js [hunk 9]
```diff
@@ -1,19 +1,55 @@
...
+    return crypto.scryptSync(plainPassword, salt, 64).toString('hex')
   }
-  return User
-}
+
+  User.getRandomSalt = function (bytes = 16) {
+    return crypto.randomBytes(bytes).toString("hex");
+  };
+
+  User.encryptPassword = function (plainPassword, salt) {
```
Comment (roke77) [inline models/user.js:45]: If you are using `getEncryptedPassword`, this `encryptPassword` is no longer useful, it does the same thing, let's remove it.

File: models/user.js [hunk 10]
```diff
@@ -1,19 +1,55 @@
...
+    console.log("SALT",salt)
```
Comment (roke77) [inline models/user.js:46]: Don't forget to clean all your console.logs

File: src/index.js [hunk 11]
```diff
@@ -16,7 +19,37 @@ const server = new ApolloServer({
...
+      const userFound = await models.User.findOne({ where: { id: jwt_payload.id } })
+
+      if (!userFound) {
+        return done(null, false)
+      }
+
+      done(null, userFound)
+    } catch (error) {
+      done(error, false)
+    }
```
Comment (roke77) [inline src/index.js:41]: There is no need of an IF statement if you are already returning what ever the `models.User.findOne` result is, so let's change shrink this code a bit:

```suggestion
    try {
      done(null, await models.User.findOne({ where: { id: jwt_payload.id } }))
    } catch (error) {
      done(error)
    }
```

File: controllers/userController.js (hunk 5 above)
Comment (NicoBellanich) [reply controllers/userController.js:41]: Sorry you mean like this:
+   });
+  return {
Or like this:
+  });
+
+  return {

File: controllers/userController.js (hunk 6 above)
Comment (NicoBellanich) [reply controllers/userController.js:61]: My bad, I pasted the function as parameter but I didn't  review the hole function for a refactoring! That's something I have to work on!

**Pull request merged**